### DISCUSSION TITLE
如何优雅地使用 KVO · /kvocontroller

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#259] [Announcements]
https://draveness.me/kvocontroller

**Comment by iTofu** [comment]
分析得很棒，赞一个👍

**Comment by hyice** [comment]
```
Fizz *fizz = [[Fizz alloc] init];
[fizz addObserver:self
       forKeyPath:@"number"
          options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
          context:nil];
fizz.number = @2;
```

我试了一下上面这段代码，发现在 removeObserver 前释放被观察者的话，好像不会 crash 了。请问是我测试方法存在什么问题，还是苹果已经修复这个问题了？

**Comment by iTofu** [comment]
@hyice 
kvo crash 的情况是观察者已经释放了，释放前没有 removeObserver，这样当被观察者的被观察属性变化时，kvo “通知”已经被释放的观察者，引起 crash。这不应该称作是一个需要修复的 bug。另外如果释放被观察者而非观察者，逻辑上也不会引起 crash。

**Comment by hyice** [comment]
@iTofu

前一种情况没有什么问题，后一种情况值得深究一下。如果先释放的是被观察者而非观察者，是不是不 removeObserver 也不会产生任何问题呢？

被观察者身上有一个`observationInfo`的属性，储存了观察者相关的信息，理论上在自己`- (void)dealloc`方法中清理掉这个属性就可以了，但我在[stackoverflow的这个问题](https://stackoverflow.com/questions/19514450/do-i-have-to-removeobserver-in-kvo-manually)中看到说这个属性会发生泄漏，实测却没有。

而这篇文章中所说的`NSInternalInconsistencyException`也是在被观察先释放的情况下触发的，我实测下来也没有遇到这个问题：

```
2017-02-26 23:44:19.720 KVOTest[15888:513229] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'An instance 0x60800001dd20 of class Fizz was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x60800003d320> (
<NSKeyValueObservance 0x608000057310: Observer: 0x7fa098f07590, Key path: number, Options: <New: YES, Old: YES, Prior: NO> Context: 0x0, Property: 0x608000057400>
)'
```

所以我才会有这个疑惑，`如果先释放的是被观察者而非观察者，是不是不 removeObserver 也不会产生任何问题呢？`

问题本身影响不大，但就是想深究一下，谢谢！

**Comment by iTofu** [comment]
@hyice 
我认为哈，Stack Overflow 这位仁兄是在扯蛋（2333，这个答案看起来就有点不合理。
A 在内部 observing 了 B，所以 A 在 dealloc 时 removeObserve，挺像“谁申请，谁释放”，这个逻辑没毛病。
但假如 A 在内部 observing 了 B，B 在 dealloc 时还要 remove A，A 申请，B 释放，这种实现方案除非苹果工程师脑子瓦特了吧。
文档也讲，A observing B，B **不会强引用** A，否则如果 A 强引用了 B，比如 B 是 A strong 引用的全局变量，这就是一个赤裸裸的循环引用嘛。
只要不是循环强引用，又怎么会造成泄漏呢，该销毁的时候总会被销毁，不用 care。

**Comment by iTofu** [comment]
@hyice 
而且我认为题主问的是 A observing B，A dealloc 时是否需要调用 removeObserver，而答案一却论证了半天 B dealloc 时是否需要调用 removeObserver。

**Comment by hyice** [comment]
@iTofu

我觉得我的理解跟你的理解基本是一致的，我也觉得B先释放的情况不应该存在问题。

不过你还是没有解决我的疑问，就是文章中的这个报错为什么会存在的问题：

```
An instance 0x60800001dd20 of class Fizz was deallocated while key value observers were still registered with it
```

Fizz就是那个被观察对象，报错中说Fizz释放的时候还有对象在观察它的变化，我觉得这个报错本身是不应该存在的，而且我没法重现这个报错，所以比较疑惑。

**Comment by iTofu** [comment]
突然发现文章我漏看了一段……这个我没实际测试过，你测试过没问题的话，要么就是这儿确实没问题，要么就是以前是有的，后来苹果更新了这块逻辑？

hyice <notifications@github.com>于2018年3月18日 周日下午10:03写道：

**Comment by hyice** [comment]
@iTofu 

我怀疑是苹果更新了这块逻辑，所以才在最开始的评论里问了博主这个问题，因为我在别的`KVO`相关文章里也看到过类似的报错，而且记得以前开发中似乎也遇到过类似的问题。不过现在我的确无法重现这个问题了。

不管实际上是怎么样的，至少从功能设计上来说，这里不抛任何异常才是比较好的选择吧。

**Comment by iTofu** [comment]
@hyice 
想起来比较常见的 kvo 场景是视频播放时监听 AVPlayerItemStatus & AVPlayerStatus，如果 observed 对象 dealloc 时确实要 removeObserve，因为 AVPlayerItem & AVPlayer 是系统类，那么系统内部需要在 dealloc 时遍历 observationInfo，去 remove 各个 observe。而自定义类的话，应该需要手动写一个这个实现，一般这样的东西类比 NSCoding、NSCopy，都会声明一个协议来交于开发者实现，但是 kvo 好像就没有，值得一究。

**Comment by iTofu** [comment]
@hyice 
既不声明相关协议交由开发者实现，系统内部也不去实现……这种逻辑真没道理

**Comment by ZenonHuang** [comment]
> @hyice
> ```
> Fizz *fizz = [[Fizz alloc] init];
> [fizz addObserver:self
>        forKeyPath:@"number"
>           options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
>           context:nil];
> fizz.number = @2;
> ```
> 
> 我试了一下上面这段代码，发现在 removeObserver 前释放被观察者的话，好像不会 crash 了。请问是我测试方法存在什么问题，还是苹果已经修复这个问题了？

------

我也特意新建了工程，发现并没有 crash。感觉博主应该更新一下文章了。

**Comment by status302** [comment]
> @ZenonHuang
> > @hyice
> > ```
> > Fizz *fizz = [[Fizz alloc] init];
> > [fizz addObserver:self
> >        forKeyPath:@"number"
> >           options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
> >           context:nil];
> > fizz.number = @2;
> > ```
> > 
> > 我试了一下上面这段代码，发现在 removeObserver 前释放被观察者的话，好像不会 crash 了。请问是我测试方法存在什么问题，还是苹果已经修复这个问题了？
> 
> ------
> 
> 我也特意新建了工程，发现并没有 crash。感觉博主应该更新一下文章了。
> 

只是在 `iOS 11` 上不会 Crash, 在 `iOS 10, 9, 8` 还是会 Crash

**Comment by Cspdhsg** [comment]
> @hyice
> @iTofu 
> 
> 我怀疑是苹果更新了这块逻辑，所以才在最开始的评论里问了博主这个问题，因为我在别的`KVO`相关文章里也看到过类似的报错，而且记得以前开发中似乎也遇到过类似的问题。不过现在我的确无法重现这个问题了。
> 
> 不管实际上是怎么样的，至少从功能设计上来说，这里不抛任何异常才是比较好的选择吧。

iOS10之后修复了这个问题
